File: backend/app/schemas/application.py
"""Application / match API schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ApplicationStatus


class ApplicationCompareRequest(BaseModel):
    job_id: uuid.UUID
    resume_id: uuid.UUID


class ApplicationApplyRequest(BaseModel):
    """Candidate self-serve apply to an open job."""

    job_id: uuid.UUID
    resume_id: uuid.UUID | None = None


class ApplicationStatusUpdate(BaseModel):
    status: ApplicationStatus


class ApplicationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    job_id: uuid.UUID
    job_title: str | None = None
    company_name: str | None = None
    candidate_id: uuid.UUID
    candidate_name: str | None = None
    candidate_email: str | None = None
    resume_id: uuid.UUID | None = None
    resume_file_name: str | None = None
    status: ApplicationStatus
    match_score: float | None = None
    ats_score: float | None = None
    matching_skills: list[str] = Field(default_factory=list)
    missing_skills: list[str] = Field(default_factory=list)
    strengths: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)
    suggestions: list[str] = Field(default_factory=list)
    summary: str | None = None
    reasoning: str | None = None
    scoring_engine: Literal["openai", "local"] | None = None
    confidence: float | None = None
    created_at: datetime
    updated_at: datetime


class ApplicationListResponse(BaseModel):
    items: list[ApplicationResponse]
    total: int
    page: int
    page_size: int
    pages: int


ApplicationSortField = Literal["created_at", "match_score", "status"]
SortOrder = Literal["asc", "desc"]

File: frontend/src/types/application.ts
export type ApplicationStatus =
  | "applied"
  | "screening"
  | "shortlisted"
  | "interview"
  | "interview_completed"
  | "offered"
  | "selected"
  | "rejected"
  | "hired"
  | "withdrawn"

export type ApplicationMatch = {
  id: string
  job_id: string
  job_title: string | null
  company_name: string | null
  candidate_id: string
  candidate_name: string | null
  candidate_email: string | null
  resume_id: string | null
  resume_file_name: string | null
  status: ApplicationStatus
  match_score: number | null
  ats_score: number | null
  matching_skills: string[]
  missing_skills: string[]
  strengths: string[]
  weaknesses: string[]
  suggestions: string[]
  summary: string | null
  reasoning: string | null
  scoring_engine?: "openai" | "local" | null
  confidence?: number | null
  created_at: string
  updated_at: string
}

export type ApplicationListResponse = {
  items: ApplicationMatch[]
  total: number
  page: number
  page_size: number
  pages: number
}

export type ComparePayload = {
  job_id: string
  resume_id: string
}

/** Re-export shared labels so existing imports keep working. */
export {
  APPLICATION_STATUS_LABELS,
  applicationStatusLabel,
} from "@/lib/application-status"
